### DISCUSSION TITLE
@can and auth()->user()->can('permission name') not working

### DISCUSSION THREAD

**Discussion by mehdisadooni** [spatie/laravel-permission#1955] [Q&A]
hi guys,
@can directive and can() method not working permission via role,
user has role and role has permissions in  my app and i want check permissions via roles not roles.
Does not check the direct  permissions and permissions via role.
but $user->hasAnyPermmission() method is working good.

what i can do to fix it?!

**Comment by erikn69** [comment]
https://github.com/spatie/laravel-permission/blob/08cc9bdbf89640cb1d64fb01eb9513c1cbb2549d/tests/GateTest.php#L40-L53

**Comment by awebartisan** [comment]
I have migrated the permissions from another package to this package. So I have roles and their permissions. No direct permissions to the User. 

Looked at the code inside `HasPermissions` trait and `hasPermissionTo()` method:
`return $this->hasDirectPermission($permission) || $this->hasPermissionViaRole($permission);` 
It first checks for direct permission and then role permissions. 

Inside `hasDirectPermission` it has this line in the end:
`return $this->permissions->contains($permission->getKeyName(), $permission->getKey());`

When we call `@can` or `can()` method on User instance, `$this->permissions` is null and throws following error:
`Call to a member function contains() on null`

I am sure I am missing something. Just can't see it. Any thoughts on this @erikn69 ?

**Comment by awebartisan** [comment]
Never mind. There was a property on User model. Dang it.

**Reply by erikn69** [reply]
[must not have a permission or permissions property nor a permissions method](https://spatie.be/index.php/docs/laravel-permission/v5/prerequisites#must-not-have-a-permission-or-permissions-property-nor-a-permissions-method)
